packages/backend/src/models/SearchModel/index.ts

```diff
@@ -326,22 +329,38 @@ export class SearchModel {
             filters?.type,
         );
 
-        const lowerCaseQuery = query.toLowerCase();
-        return explores
+        // Building regex to match any of the words in the query and then using it to match against the label and description
+        // results are sorted by the number of matches - we create a set out of the matches to remove duplicates
+        const sanitizedQuery = escapeRegExp(query);
+        const splitQuery = compact(
+            Array.from(new Set(sanitizedQuery.split(' '))),
+        );
+
+        const splitQueryRegex = new RegExp(splitQuery.join('|'), 'ig');
```

Comment:
Since we are trying to find the most relevant matches here, it might be beneficial to use Fuse.js, which we already utilize in the frontend and does not incur a performance overhead. It also offers a way to return relevance scores and features a full-text-like matching algorithm, which is even better because it works great with substrings

Reply:
After some testing with fuse, we can see an impact on the time it takes to search the fields and tables. This is mostly because we first need to fetch all the results in the explore so that then we can use `fuse` to perform the search.

It can will also get more expensive as the number of `fields` and `tables` grow.
This isn't the case with the `regex` approach because we stop collecting results as soon as the limit is reached - ultimately can produce worse results

Here's the file with the collected times. I added a model with 5000 columns and tried it out with both `fuse` and previous `regex` implementation.

After collecting these, I also tried to add a model with 10000 columns, and could see that `fuse` was even slower, but `regex` remained around the same time.

[times.md](https://github.com/lightdash/lightdash/files/14423050/times.md)

Reply:
We ended up deciding going the `regex` route, first collecting all the `fields` and `tables` that matched it and then getting the top 10 results - to have more complete results. This still ended up being faster than fuse (tested with model having 40000 columns):

- `{ sortedTablesLen: 1, sortedFieldsLen: 40000 }`
- `searchTablesAndFields: 20-30ms`